Implementing NpSigninDialog and Updating NP Library.

anr2me opened the pull request:
This were a part of my piled uncommitted infrastructure codes, just trying to create a separate PR to reduce the pile.

PS: I think i need to move some of the registered module(s) on sceNp.cpp into a separate files (ie. sceNp2.cpp), similar to what RPCS3 did, since i'm adding new dialog files anyway might as well create the sceNp2 files.

anr2me committed:
Implement NP Signin Dialog (faked signin)

anr2me committed:
Updated NP Library
PS: Some of registered modules should probably moved to a separate files (ie. sceNp2.cpp).

anr2me committed:
Added a FunctionWrapper used by sceNpService

hrydgard commented:
CI is finding an error:
```
/home/runner/work/ppsspp/ppsspp/Core/HLE/sceNet.cpp:433:25: error: cannot bind packed field ‘args.NpAuthArgs::data[0]’ to ‘unsigned int&’
  433 |   auto& id = args.data[0];
```

Hm, maybe type inference issue? Try declaring the type directly instead of using auto&, maybe...

ANR2ME commented:
> CI is finding an error:
> 
> ```
> /home/runner/work/ppsspp/ppsspp/Core/HLE/sceNet.cpp:433:25: error: cannot bind packed field ‘args.NpAuthArgs::data[0]’ to ‘unsigned int&’
>   433 |   auto& id = args.data[0];
> ```
> 
> Hm, maybe type inference issue? Try declaring the type directly instead of using auto&, maybe...

yeah, referencing a packed data such as array is a possible alignment issue, since the compiler no longer know whether it was packed or not when using the reference.

i'll probably remove that `&` on another PR later, just to be safe.

unknownbrackets commented on Core/HLE/sceNp.h:
```diff
@@ -17,6 +17,17 @@
+#include "Common/Net/Resolve.h"
+#include "Common/Net/URL.h"
+
+#ifdef _MSC_VER
+#define PACK  // on MSVC we use #pragma pack() instead so let's kill this.
+#else
+#define PACK __attribute__((packed))
+#endif
+
 #ifdef _MSC_VER
 #pragma pack(push,1)
 #endif
```
FWIW, now in other places we simply use `#pragma pack(push, 1)` on all compilers.  See Core/Replay.cpp or GPU/Debugger/RecordFormat.h.  We don't really support ancient GCC versions anyway, so everything we compile on supports `#pragma pack(push, 1)`.

It'd be nice if there was some fancy c++25 way of doing this or something.

But anyway, this would allow you to avoid the ugly C-style struct typedefs.

-[Unknown]

ANR2ME replied:
i see, i thought `#pragma pack(push, 1)` only worked for MSVC.

ANR2ME commented on Core/Dialog/PSPNpSigninDialog.cpp:
```diff
@@ -0,0 +1,349 @@
+}
+
+void PSPNpSigninDialog::DrawLogo() {
+	// TODO: Draw DNAS logo
```
it's just a stub for now showing an empty box, since there is no logo on the atlas, but we can probably replace the official DNAS logo with OpenDNAS logo

anr2me committed:
Moving sceNpMatching2-stuff to sceNp2.cpp/h.
PS: Register_sceNpMatching2 not being added to HLETables because currently the callback didn't work properly yet and could crash the game.

anr2me committed:
Replace the dummy PSN name with Nickname on sceNpInit.

MojoJojoDojo commented:
So this should be merged now?

hrydgard commented:
I've only skimmed it, but looks OK to me. What do you say @ANR2ME , ready to go?

ANR2ME commented:
Yes, it's ready.
I'm planning to create PR for infrastructure progressively (may be per library), so this PR need to be merged first before i can create the next PR.

unknownbrackets commented on Core/Dialog/PSPNpSigninDialog.cpp:
```diff
@@ -0,0 +1,349 @@
+#if defined(_WIN32)
+#include "Common/CommonWindows.h"
+#endif
+#include <TimeUtil.h>
```
```suggestion
#include "Common/TimeUtil.h"
```

Use `<>` for system headers that come from the OS or external installed libraries, and use `""` for headers that are part of the source code.  `Common` is part of the source code.  In theory, some OS like OS/2 or Windows 12 could have a TimeUtil.h header and `<>` would prefer the system header over the correct one.

-[Unknown]

unknownbrackets commented on Core/HLE/sceNp.cpp:
```diff
@@ -18,28 +18,32 @@
 // to keep games happy anyway.
 
-#include <mutex>
-#include <deque>
```
nit: Since this file uses a deque, I recommend including it in the cpp file.  Even if it's also needed in an h file, the check to include once is fairly cheap.  It makes it easier later to prune down unnecessary headers in `.h` files, which can really improve overall compile time.

-[Unknown]

unknownbrackets commented on Core/HLE/sceNp.cpp:
```diff
@@ -18,28 +18,32 @@
 #include <StringUtils.h>
 #include "Core/MemMapHelpers.h"
 #include <Core/CoreTiming.h>
+#include <Core/Config.h>
```
Should not use `<>`, see other comment.

-[Unknown]

unknownbrackets commented on Core/HLE/sceNp.h:
```diff
@@ -17,9 +17,16 @@
+#include <mutex>
+#include <deque>
+#include <map>
+#include <Core/HLE/sceRtc.h>
```
nit: Should not be `<>`.

-[Unknown]

unknownbrackets commented on Core/HLE/sceNp2.cpp:
```diff
@@ -0,0 +1,566 @@
+// Copyright (c) 2012- PPSSPP Project.
```
nit: New files should ideally use current year (i.e. `2022-`.)  I don't know of any locality where this would actually matter, though, but I'm not a licensed law person...

-[Unknown]

unknownbrackets commented on Core/HLE/sceNp2.cpp:
```diff
@@ -0,0 +1,566 @@
+// This is pretty much a stub implementation. Doesn't actually do anything, just tries to return values
+// to keep games happy anyway.
+
+#include <StringUtils.h>
+#include "Core/MemMapHelpers.h"
+#include <Core/CoreTiming.h>
```
Should use `""`.  Neither of these are system headers from the OS or installed libraries, we want our include path to take precedence.

-[Unknown]

unknownbrackets reviewed (commented):
Took a quick look at the Np files which I didn't really look at before.

-[Unknown]

ANR2ME commented:
Looks like i'll need to revise it again :) Thanks unknownbrackets i didn't aware the difference between `<>` and `""` before.

anr2me committed:
Updated copyright's year.

anr2me committed:
Fix include issue.

ANR2ME commented:
Will this do?
